File: agentic-programming-scratch/ising.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <math.h>

static int N;
static int8_t *lattice;

static inline int idx(int i, int j) {
    return i * N + j;
}

static inline double uniform(void) {
    return rand() / (RAND_MAX + 1.0);
}

static void init_lattice(unsigned int seed) {
    srand(seed);
    for (int i = 0; i < N * N; i++) {
        lattice[i] = (int8_t) ((rand() % 2) * 2 - 1);
    }
}

static void metropolis_sweep(double beta) {
    for (int k = 0; k < N * N; k++) {
        int i = rand() % N;
        int j = rand() % N;
        int s = lattice[idx(i, j)];
        int neighbor_sum = lattice[idx((i + N - 1) % N, j)] + lattice[idx((i + 1) % N, j)]
                          + lattice[idx(i, (j + N - 1) % N)] + lattice[idx(i, (j + 1) % N)];
        double dE = 2.0 * s * neighbor_sum;
        if (dE <= 0.0 || uniform() < exp(-beta * dE)) {
            lattice[idx(i, j)] = (int8_t) -s;
        }
    }
}

static double total_energy(void) {
    double e = 0.0;
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            int s = lattice[idx(i, j)];
            e -= s * (lattice[idx(i, (j + 1) % N)] + lattice[idx((i + 1) % N, j)]);
        }
    }
    return e;
}

static long magnetization(void) {
    long m = 0;
    for (int i = 0; i < N * N; i++) m += lattice[i];
    return m;
}

int main(void) {
    N = 32;
    double T = 2.269;
    int n_equil = 1000;
    int n_sweeps = 2000;

    lattice = malloc((size_t) N * N * sizeof(int8_t));
    init_lattice(0);

    double beta = 1.0 / T;

    for (int s = 0; s < n_equil; s++) {
        metropolis_sweep(beta);
    }

    double e_sum = 0.0, m_sum = 0.0;
    for (int s = 0; s < n_sweeps; s++) {
        metropolis_sweep(beta);
        e_sum += total_energy();
        m_sum += fabs((double) magnetization());
    }

    int n_spins = N * N;
    printf("N = %dx%d, T = %.3f\n", N, N, T);
    printf("<E>/spin = %.4f\n", e_sum / n_sweeps / n_spins);
    printf("<|M|>/spin = %.4f\n", m_sum / n_sweeps / n_spins);

    free(lattice);
    return 0;
}

File: agentic-programming-scratch/ising.py
import numpy as np


def random_lattice(n, rng):
    return rng.choice(np.array([-1, 1], dtype=np.int8), size=(n, n))


def _neighbor_sum(lattice):
    return (
        np.roll(lattice, 1, axis=0) + np.roll(lattice, -1, axis=0)
        + np.roll(lattice, 1, axis=1) + np.roll(lattice, -1, axis=1)
    )


def metropolis_sweep(lattice, beta, checkerboard, rng):
    """One sweep = both checkerboard colors updated once (Metropolis, periodic BCs).

    Same-color sites share no bond, so a color's spins can be flipped
    simultaneously from a single vectorized accept/reject draw.
    """
    for color in (0, 1):
        mask = checkerboard == color
        dE = 2.0 * lattice * _neighbor_sum(lattice)
        accept = (dE <= 0.0) | (rng.random(lattice.shape) < np.exp(-beta * dE))
        flip = mask & accept
        lattice[flip] *= -1
    return lattice


def total_energy(lattice):
    # Only right+down neighbors: each bond <ij> counted exactly once.
    return -np.sum(lattice * (np.roll(lattice, -1, axis=0) + np.roll(lattice, -1, axis=1)))


def magnetization(lattice):
    return lattice.sum()


def run(n=32, temperature=2.269, n_equil_sweeps=1000, n_sweeps=2000, seed=0):
    """Equilibrate then sample. Returns final lattice and per-sweep E, M series."""
    rng = np.random.default_rng(seed)
    lattice = random_lattice(n, rng)
    beta = 1.0 / temperature
    checkerboard = np.indices((n, n)).sum(axis=0) % 2

    for _ in range(n_equil_sweeps):
        metropolis_sweep(lattice, beta, checkerboard, rng)

    energies = np.empty(n_sweeps)
    mags = np.empty(n_sweeps)
    for k in range(n_sweeps):
        metropolis_sweep(lattice, beta, checkerboard, rng)
        energies[k] = total_energy(lattice)
        mags[k] = magnetization(lattice)

    return lattice, energies, mags


if __name__ == "__main__":
    N = 32
    T = 2.269  # ~ critical temperature for the 2D Ising model, J = k_B = 1
    lattice, energies, mags = run(n=N, temperature=T)

    n_spins = N * N
    print(f"N = {N}x{N}, T = {T}")
    print(f"<E>/spin = {energies.mean() / n_spins:.4f}")
    print(f"<|M|>/spin = {np.abs(mags).mean() / n_spins:.4f}")
